ui_test/src/comments.rs

```diff
@@ -140,36 +208,141 @@ impl Comments {
+        checked!(path, l);
+        match command {
+            "revisions" => {
+                check!(self.revisions.is_none(), "cannot specifiy revisions twice");
+                self.revisions = Some(args.split_whitespace().map(|s| s.to_string()).collect());
+            }
+            "compile-flags" => {
+                self.compile_flags.extend(args.split_whitespace().map(|s| s.to_string()));
+            }
+            "rustc-env" =>
+                for env in args.split_whitespace() {
+                    let (k, v) = unwrap!(
+                        env.split_once('='),
+                        "environment variables must be key/value pairs separated by a `=`"
+                    );
+                    self.env_vars.push((k.to_string(), v.to_string()));
+                },
+            "normalize-stderr-test" => {
+                let (from, to) =
+                    unwrap!(args.split_once("->"), "normalize-stderr-test needs a `->`");
+                let from = from.trim().trim_matches('"');
+                let to = to.trim().trim_matches('"');
+                let from = unwrap!(Regex::new(from).ok(), "invalid regex");
+                self.normalize_stderr.push((from, to.to_string()));
+            }
+            "error-pattern" => {
+                check!(
+                    self.error_pattern.is_none(),
+                    "cannot specifiy error_pattern twice, previous: {:?}",
+                    self.error_pattern
+                );
+                self.error_pattern = Some((args.trim().to_string(), l));
+            }
+            // Maybe the user just left a comment explaining a command without arguments
+            _ => self.parse_command(command, path, l),
+        }
+    }
 
-                let level = captures.name("level").map(|rev| rev.as_str().parse().unwrap());
+    fn parse_command(&mut self, command: &str, path: &Path, l: usize) {
+        checked!(path, l);
 
-                let match_line = match captures.name("offset").map(|rev| rev.as_str()) {
-                    Some("|") => fallthrough_to.expect("`//~|` pattern without preceding line"),
-                    Some(pat) => {
-                        debug_assert!(pat.chars().all(|c| c == '^'));
-                        l - pat.len()
-                    }
-                    None => l,
-                };
-
-                fallthrough_to = Some(match_line);
-
-                this.error_matches.push(ErrorMatch {
-                    matched,
-                    revision,
-                    level,
-                    definition_line: l,
-                    line: match_line,
-                });
-            } else {
-                fallthrough_to = None;
-            }
+        if let Some(s) = command.strip_prefix("ignore-") {
+            self.ignore.push(Condition::parse(s));
+            return;
         }
-        this
+
+        if let Some(s) = command.strip_prefix("only-") {
+            self.only.push(Condition::parse(s));
```

Comment:
Note that `Condition::parse` can panic... now that you have error reporting here it'd be nice to also use that for incorrect conditions.

For proper reliable parsing, I think `ignore-windows` should become `ignore-target-windows`, otherwise we have the same problem again of having some special commands (ignore-debug, ignore-32bit) and then silently falling back to another way of interpreting the command.

Reply:
agreed. For minimal churn I want to change as little as possible until we rip out the old logic and then improve the parser step by step.